code/controllers/subsystem/processing/overlays.dm

```diff
@@ -0,0 +1,160 @@
+		var/list/cached_icon = icon_states_cache[icon]
+		if (cached_icon)
+			var/cached_appearance = cached_icon["[image_atom_or_string]"]
+			if (cached_appearance)
+				return cached_appearance
+		stringbro.icon = icon
+		stringbro.icon_state = image_atom_or_string
+		if (!cached_icon) //not using the macro to save an associated lookup
+			cached_icon = list()
+			icon_states_cache[icon] = cached_icon
+		var/cached_appearance = stringbro.appearance
+		cached_icon["[image_atom_or_string]"] = cached_appearance
+		return cached_appearance
+	else if (istype(image_atom_or_string, /image) || istype(image_atom_or_string, /atom))
+		var/image/I = image_atom_or_string
+		return I.appearance
+#if DM_VERSION >= 511
+	else if (istype(image_atom_or_string, /mutable_appearance/))
+		var/mutable_appearance/MA = image_atom_or_string
+		return MA.appearance
+#endif
+
+#define NOT_QUEUED_ALREADY (!(flags & OVERLAY_QUEUED))
+#define QUEUE_FOR_COMPILE flags |= OVERLAY_QUEUED; SSoverlays.can_fire = TRUE; if(SSoverlays.initialized) { SSoverlays.processing += src; } else { SSoverlays.processing[src] = src; } 
+/atom/proc/cut_overlays(priority = FALSE)
+	var/list/cached_overlays = our_overlays
+	var/list/cached_priority = priority_overlays
+	
+	var/need_compile = FALSE
+
+	if(LAZYLEN(cached_overlays)) //don't queue empty lists, don't cut priority overlays
+		cached_overlays.Cut()  //clear regular overlays
+		need_compile = TRUE
+
+	if(priority && LAZYLEN(cached_priority))
+		cached_priority.Cut()
+		need_compile = TRUE
+
+	if(NOT_QUEUED_ALREADY && need_compile)
+		QUEUE_FOR_COMPILE
+
+/atom/proc/cut_overlay(image, priority)
+	if(!image)
+		return
+	var/list/cached_overlays = our_overlays	//sanic
+	var/list/cached_priority = priority_overlays
+	var/init_o_len = LAZYLEN(cached_overlays)
+	var/init_p_len = LAZYLEN(cached_priority)  //starter pokemon
+	
+	image = extract_overlay_appearance(image)
+
+	LAZYREMOVE(cached_overlays, image)
+	if(priority)
+		LAZYREMOVE(cached_priority, image)
+
+	if(NOT_QUEUED_ALREADY && ((init_o_len != LAZYLEN(cached_priority)) || (init_p_len != LAZYLEN(cached_overlays))))
+		QUEUE_FOR_COMPILE
+
+/atom/proc/add_overlay(image, priority = FALSE)
+	if(!image)
+		return
+	if(islist(image))
+		for(var/I in image)
+			add_overlay(I)
+	LAZYINITLIST(our_overlays)	//always initialized after this point
+	LAZYINITLIST(priority_overlays)
+
+	image = extract_overlay_appearance(image)
+
+	var/list/cached_overlays = our_overlays	//sanic
+	var/list/cached_priority = priority_overlays
+	var/init_o_len = cached_overlays.len
+	var/init_p_len = cached_priority.len  //starter pokemon
+	var/need_compile
+
+	if(priority)
+		cached_priority += image  //or in the image. Can we use [image] = image?
+		need_compile = init_p_len != cached_priority.len
+	else
+		cached_overlays += image
+		need_compile = init_o_len != cached_overlays.len
+
+	if(NOT_QUEUED_ALREADY && need_compile) //have we caught more pokemon?
+		QUEUE_FOR_COMPILE
```

Comment:
I don't think you should call these image considering they're not images and have absolutely nothing to do with images.